carma_wm/src/SignalizedIntersectionManager.cpp

```diff
@@ -0,0 +1,325 @@
+  {
+    target_frame_ = target_frame;
+  }
+
+  void SignalizedIntersectionManager::convertLaneToLaneletId(std::unordered_map<uint8_t, lanelet::Id>& entry, std::unordered_map<uint8_t, lanelet::Id>& exit, const cav_msgs::IntersectionGeometry& intersection, const std::shared_ptr<lanelet::LaneletMap>& map)
+  {
+    std::unordered_map<uint8_t, std::unordered_set<uint16_t>> signal_group_to_exit_lanes;
+    std::unordered_map<uint8_t, std::unordered_set<uint16_t>> signal_group_to_entry_lanes;
+
+    if (target_frame_ == "")
+    {
+      throw std::invalid_argument("Map is not initialized yet as the georeference was not found...");
+    }
+
+    lanelet::projection::LocalFrameProjector local_projector(target_frame_.c_str());
+
+    lanelet::GPSPoint gps_point;
+    gps_point.lat = intersection.ref_point.latitude;
+    gps_point.lon = intersection.ref_point.longitude;
+    gps_point.ele = intersection.ref_point.elevation;
+
+    auto ref_node  = local_projector.forward(gps_point);
+
+    ROS_DEBUG_STREAM("Reference node in map frame x: " << ref_node.x() << ", y: " << ref_node.y());
+    
+    std::vector<lanelet::BasicPoint2d> node_list;
+
+    for (auto lane : intersection.lane_list)
+    {
+      lanelet::BasicPoint2d curr_node = {ref_node.x(), ref_node.y()};
+      ROS_DEBUG_STREAM("Processing Lane id: " << (int)lane.lane_id);
+      
+      size_t min_number_of_points = 2; // only two points are sufficient to get corresponding lanelets
+
+      for (size_t i = 0; i < min_number_of_points; i ++ )
+      {
+        curr_node.x() = lane.node_list.nodes.node_set_xy[i].delta.x + curr_node.x();
+        curr_node.y() = lane.node_list.nodes.node_set_xy[i].delta.y + curr_node.y();
+        ROS_DEBUG_STREAM("Current node x: " << curr_node.x() << ", y: " << curr_node.y());
+
+        node_list.push_back(curr_node);
+      }
+
+      // save which signal group connect to which exit lanes
+      for (auto connection : lane.connect_to_list)
+      {
+        signal_group_to_exit_lanes[connection.signal_group].emplace(connection.connecting_lane.lane);
+        
+        if (lane.lane_attributes.directional_use.lane_direction == LANE_DIRECTION::INGRESS)
+          signal_group_to_entry_lanes[connection.signal_group].emplace(lane.lane_id);
+      }
+      
+      // query corresonding lanelet lane from local map
+      auto first_llts = carma_wm::query::getLaneletsFromPoint(map, node_list.front());
+      auto second_llts = carma_wm::query::getLaneletsFromPoint(map, node_list.back());
+
+      if (first_llts.empty() && second_llts.empty())
+      {
+        throw std::invalid_argument("Given offset points are not inside the map...");
+      }
+
+      lanelet::Id corresponding_lanelet_id = second_llts.front().id(); // by default second point's lanelet as first
+                                                                        // point may fall exactly on boundaries of two llts.
+      for (auto i : second_llts)
+      {
+        for (auto j : first_llts)
```

Comment:
Good thinking. This logic needs an improvement and I am working on it right now. Initial thought is to import WMBroadcaster's getAffectedLaneletOrAreas function into a utility library and access from there.

Reply:
I imported above mentioned function as a library now such that when the given 2 offset points are across multiple lanelets, this will pick only the first lanelet.